File: src/windows-native/LanMessenger/Core/Crypto/SessionCrypto.cs
using NSec.Cryptography;
using System.Security.Cryptography;
using System.Text;

namespace LanMessenger.Core.Crypto;

// Handles per-peer message encryption/decryption.
//
// Protocol:
//   shared_secret = X25519(my_private, peer_public)
//   symmetric_key = HKDF-SHA256(ikm: shared_secret, salt: [], info: "lan-messenger", len: 32)
//   nonce         = random(12 bytes)
//   ciphertext    = AES-256-GCM.Encrypt(key, nonce, plaintext, aad)
//   transmitted   = base64(nonce) + base64(ciphertext ‖ 16-byte tag)
//
// The 16-byte AES-GCM tag is appended to the ciphertext bytes before base64-encoding,
// matching the Python cryptography library's output format.
//
// AES-GCM uses System.Security.Cryptography.AesGcm (BCL) rather than NSec's
// Aes256Gcm so it runs on CPUs without AES-NI (e.g. older Windows 10 machines
// and VMs without hardware AES passthrough). NSec is retained only for X25519
// key agreement and HKDF key derivation, which have no AES-NI dependency.
public static class SessionCrypto
{
    private static readonly X25519     _x25519  = KeyAgreementAlgorithm.X25519;
    private static readonly HkdfSha256 _hkdf    = KeyDerivationAlgorithm.HkdfSha256;

    private static readonly byte[] _hkdfInfo   = Encoding.UTF8.GetBytes("lan-messenger");
    private static readonly byte[] _emptyBytes = [];

    private const int KeySize   = 32;
    private const int NonceSize = 12;
    private const int TagSize   = 16;

    // Derive the 32-byte shared symmetric key for a peer.
    public static byte[] SymmetricKey(Key myPrivateKey, string peerPublicKeyB64)
    {
        byte[] peerRaw = Convert.FromBase64String(peerPublicKeyB64);
        if (peerRaw.Length != 32) throw new ArgumentException("Public key must be 32 bytes");

        var peerPub = PublicKey.Import(_x25519, peerRaw, KeyBlobFormat.RawPublicKey);
        var sharedSecret = _x25519.Agree(myPrivateKey, peerPub)
            ?? throw new CryptographicException("X25519 agreement returned null");

        return _hkdf.DeriveBytes(sharedSecret, _emptyBytes, _hkdfInfo, KeySize);
    }

    // Encrypt. Returns (nonceB64, ciphertextB64) where ciphertextB64 encodes ciphertext ‖ 16-byte tag.
    public static (string NonceB64, string CiphertextB64) Encrypt(byte[] keyBytes, byte[] plaintext, byte[] aad)
    {
        byte[] nonce      = new byte[NonceSize];
        byte[] ciphertext = new byte[plaintext.Length];
        byte[] tag        = new byte[TagSize];
        RandomNumberGenerator.Fill(nonce);

        using var aesGcm = new AesGcm(keyBytes, TagSize);
        aesGcm.Encrypt(nonce, plaintext, ciphertext, tag, aad);

        byte[] ciphertextWithTag = new byte[ciphertext.Length + TagSize];
        ciphertext.CopyTo(ciphertextWithTag, 0);
        tag.CopyTo(ciphertextWithTag, ciphertext.Length);

        return (Convert.ToBase64String(nonce), Convert.ToBase64String(ciphertextWithTag));
    }

    // Decrypt. ciphertextB64 must include the 16-byte tag appended.
    public static byte[] Decrypt(byte[] keyBytes, string nonceB64, string ciphertextB64, byte[] aad)
    {
        byte[] nonce = Convert.FromBase64String(nonceB64);
        if (nonce.Length != NonceSize) throw new CryptographicException("Nonce must be 12 bytes");

        byte[] ciphertextWithTag = Convert.FromBase64String(ciphertextB64);
        if (ciphertextWithTag.Length < TagSize) throw new CryptographicException("Ciphertext too short");

        int    ciphertextLen = ciphertextWithTag.Length - TagSize;
        byte[] ciphertext    = ciphertextWithTag[..ciphertextLen];
        byte[] tag           = ciphertextWithTag[ciphertextLen..];
        byte[] plaintext     = new byte[ciphertextLen];

        using var aesGcm = new AesGcm(keyBytes, TagSize);
        aesGcm.Decrypt(nonce, ciphertext, tag, plaintext, aad);
        return plaintext;
    }

    // Convenience: encrypt for a specific peer.
    public static (string NonceB64, string CiphertextB64) EncryptForPeer(
        Key myPrivateKey, string peerPublicKeyB64, byte[] plaintext, byte[] aad)
    {
        var keyBytes = SymmetricKey(myPrivateKey, peerPublicKeyB64);
        return Encrypt(keyBytes, plaintext, aad);
    }

    // Convenience: decrypt from a specific peer.
    public static byte[] DecryptFromPeer(
        Key myPrivateKey, string peerPublicKeyB64, string nonceB64, string ciphertextB64, byte[] aad)
    {
        var keyBytes = SymmetricKey(myPrivateKey, peerPublicKeyB64);
        return Decrypt(keyBytes, nonceB64, ciphertextB64, aad);
    }
}

File: src/macos/LanMessenger/Core/Crypto/SessionCrypto.swift
import Foundation
import CryptoKit

enum SessionCryptoError: Error {
    case invalidPublicKey
    case invalidNonce       // not 12 bytes
    case decryptionFailed
    case encodingFailed
}

// Handles per-peer message encryption/decryption.
//
// Protocol:
//   shared_secret = X25519(my_private, peer_public)
//   symmetric_key = HKDF-SHA256(ikm: shared_secret, salt: [], info: "lan-messenger", len: 32)
//   nonce         = random(12)
//   ciphertext    = AES-256-GCM.seal(plaintext, key, nonce, aad)
//   transmitted   = base64(nonce) + base64(ciphertext ‖ tag)
//
// The 16-byte AES-GCM tag is appended to the ciphertext bytes before base64-encoding,
// matching the Python cryptography library's output format.
enum SessionCrypto {

    // Derive the shared symmetric key from our private key and the peer's public key.
    static func symmetricKey(
        myPrivate: Curve25519.KeyAgreement.PrivateKey,
        theirPublicKeyB64: String
    ) throws -> SymmetricKey {
        guard let rawData = Data(base64Encoded: theirPublicKeyB64),
              rawData.count == 32 else {
            throw SessionCryptoError.invalidPublicKey
        }
        let theirPublic = try Curve25519.KeyAgreement.PublicKey(rawRepresentation: rawData)
        let sharedSecret = try myPrivate.sharedSecretFromKeyAgreement(with: theirPublic)
        return sharedSecret.hkdfDerivedSymmetricKey(
            using: SHA256.self,
            salt: Data(),                          // empty salt — matches Python `salt=None`
            sharedInfo: Data("lan-messenger".utf8),
            outputByteCount: 32
        )
    }

    // Encrypt plaintext. Returns (nonce_b64, ciphertext_b64).
    // ciphertext_b64 encodes `ciphertext ‖ tag` (tag appended, 16 bytes).
    static func encrypt(
        key: SymmetricKey,
        plaintext: Data,
        aad: Data
    ) throws -> (nonceB64: String, ciphertextB64: String) {
        let nonce = AES.GCM.Nonce()
        let sealed = try AES.GCM.seal(plaintext, using: key, nonce: nonce, authenticating: aad)
        let combined = sealed.ciphertext + sealed.tag      // ciphertext ‖ 16-byte tag
        return (
            Data(nonce).base64EncodedString(),
            combined.base64EncodedString()
        )
    }

    // Decrypt. nonce must be exactly 12 bytes. ciphertext includes the 16-byte tag appended.
    static func decrypt(
        key: SymmetricKey,
        nonceB64: String,
        ciphertextB64: String,
        aad: Data
    ) throws -> Data {
        guard let nonceData = Data(base64Encoded: nonceB64), nonceData.count == 12 else {
            throw SessionCryptoError.invalidNonce
        }
        guard let combined = Data(base64Encoded: ciphertextB64), combined.count >= 16 else {
            throw SessionCryptoError.decryptionFailed
        }
        let nonce = try AES.GCM.Nonce(data: nonceData)
        let ciphertext = combined.dropLast(16)
        let tag = combined.suffix(16)
        let box = try AES.GCM.SealedBox(nonce: nonce, ciphertext: ciphertext, tag: tag)
        do {
            return try AES.GCM.open(box, using: key, authenticating: aad)
        } catch {
            throw SessionCryptoError.decryptionFailed
        }
    }

    // Convenience: encrypt for a specific peer given their public key b64.
    static func encryptForPeer(
        myPrivate: Curve25519.KeyAgreement.PrivateKey,
        peerPublicKeyB64: String,
        plaintext: Data,
        aad: Data
    ) throws -> (nonceB64: String, ciphertextB64: String) {
        let key = try symmetricKey(myPrivate: myPrivate, theirPublicKeyB64: peerPublicKeyB64)
        return try encrypt(key: key, plaintext: plaintext, aad: aad)
    }

    // Convenience: decrypt from a specific peer.
    static func decryptFromPeer(
        myPrivate: Curve25519.KeyAgreement.PrivateKey,
        peerPublicKeyB64: String,
        nonceB64: String,
        ciphertextB64: String,
        aad: Data
    ) throws -> Data {
        let key = try symmetricKey(myPrivate: myPrivate, theirPublicKeyB64: peerPublicKeyB64)
        return try decrypt(key: key, nonceB64: nonceB64, ciphertextB64: ciphertextB64, aad: aad)
    }
}
